\name{IdentifyBackground}
\alias{IdentifyBackground}
\title{
  Identify annotation for background noise
}
\description{
%%  ~~ A concise (1-5 lines) description of what the function does. ~~
}
\usage{
IdentifyBackground(organism, bed_path, binsize, promo_bed, cores)
}
%- maybe also 'usage' for other objects documented here.
\arguments{
  \item{organism}{
  Accession code for organism that you would like to extract information from. For a list of availabe organisms, please see \code{\link{ShowAvailableOrg}}
  }
  \item{bed_path}{
  Absolute path you wanna to save the internally generated bed file that is created by user-specified bin size
  }
  \item{binsize}{
  Bin size in bases to create a sliding window across the genome
  }
  \item{promo_bed}{
  The promoter annotation generated by \code{\link{GetPromoterAnno}}
  }
}
\details{
  To identify regions that potentially contribute to background noise, we built a 100 bp tiling window across the whole genome. In order to identify the regions for measuring background noise, we applied following procedures:

  1. Filtering step: In all the 100 bp windows, exclude any window that resides in promoter regions, predicted CpG islands regions and windows that contains ambiguous bases (gaps)

  Firstly, construct potentially functional regions to exclude from 100bp windows
  Then, extract sequence files for the newly constructed 100bp windows. The output fasta file will be used to calcualte GC content statistics in whole genome sliding windows:
  Finally, filter 100 bp windows by excluding those that intersect with functional elements
  2. Construct preliminary background regions based on GC content: for each promoter region, identify 80 100bp windows nearby that has low in GC content (< 40%) and also relatively proximal to the corresponding TSS
}
\value{
A data.frame that contains the following elements:
\item{chrom}{
  Chromsome number
}
\item{chromStart}{
  Chromsome Start coordinate
}
\item(chromEnd){
  Chromsome End coordinate
}
\item(GC){
  GC content
}
\item{proximity}{
  Distance to cloest gene
}  
\item{geneName}{
  Cloest gene name
}
}
\references{
  https://github.com/Liuy12/MBDDiff
}
\author{
  Yuanhang Liu
}
\seealso{\code{\link{Summerizebg}}}
\examples{
\dontrun{
    bed_100bp_bg <- IdentifyBackground(organism = 'hg19', bed_path = 'path/to/bed', binsize = 100, promo_bed = Promoter_anno, cores = 4)
  }
}